### PR TITLE
Added compatibility layer API

### PR DESCRIPTION
**Pull request by miyazakh**
Added 
 SSL_CONF_cmd
 SSL_CONF_CTX_finish

 Currently supported commands:
  config. command: Curves/Certificate/PrivateKey/CipherString/DHParameters
  cmd line commands:  -curves/-cert/-key/-cipher/-dhparam

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -55969,36 +55971,410 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+            return WOLFSSL_FAILURE;
+        }
+        dh = wolfSSL_PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
+        if (!dh) {
+            wolfSSL_BIO_free(bio);
+            WOLFSSL_MSG("PEM read bio failed");
+            return WOLFSSL_FAILURE;
+        }
+    } else
+        return 1;
```
Comment (cconlon) [inline src/ssl.c:56219]: Let's put brackets around this else.

File: src/ssl.c [hunk 2]
```diff
@@ -55969,36 +55971,410 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+    
+    if (cctx->ctx) {
+        ret = (int)wolfSSL_CTX_set_tmp_dh(cctx->ctx, dh);
+    }
+    
+    if (cctx->ssl) {
```
Comment (cconlon) [inline src/ssl.c:56225]: Check "ret" value in if statement to avoid resetting by following call.

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/ssl.c [hunk 3]
```diff
@@ -55969,36 +55971,410 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+        WOLFSSL_MSG("bad arguments");
+        return ret;
+    }
+    
+    if (cctx->ctx) {
+        ret = wolfSSL_CTX_set_cipher_list(cctx->ctx, value);
+    }
+    
+    if (cctx->ssl) {
+        ret = wolfSSL_set_cipher_list(cctx->ssl, value);
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:56044]: Agree. Added check ret in the "if" first.

File: src/ssl.c [hunk 4]
```diff
@@ -55969,36 +55971,410 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+    if (value == NULL) {
+        WOLFSSL_MSG("bad arguments");
+        return ret;
+    }
+    
+    if (cctx->ctx) {
+        ret = wolfSSL_CTX_set1_curves_list(cctx->ctx, value);
+    }
+    
+    if (cctx->ssl) {
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:56080]: Agree. Added check ret in the "if" first.

File: src/ssl.c [hunk 5]
```diff
@@ -55969,36 +55971,410 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+    if (!(cctx->flags & WOLFSSL_CONF_FLAG_CERTIFICATE)) {
+        WOLFSSL_MSG("certificate flag is not set");
+        return -2;
+    }
+    
+    if (cctx->ctx) {
+        ret = wolfSSL_CTX_use_certificate_chain_file(cctx->ctx, value);
+    }
+    
+    if (cctx->ssl) {
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:56123]: Agree. Added check ret in the "if" first.

File: src/ssl.c [hunk 6]
```diff
@@ -55969,36 +55971,410 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+        WOLFSSL_MSG("certificate flag is not set");
+        return -2;
+    }
+    
+    if (cctx->ctx) {
+        ret = wolfSSL_CTX_use_PrivateKey_file(cctx->ctx, value, 
+                                                    WOLFSSL_FILETYPE_PEM);
+    }
+    
+    if (cctx->ssl) {
```
[the remark this answers is not in the document]
Comment (miyazakh) [reply src/ssl.c:56165]: Agree. Added check ret in the "if" first.

File: src/ssl.c (hunk 1 above)
Comment (miyazakh) [reply src/ssl.c:56219]: Thanks for pointing it out. Added.

File: src/ssl.c (hunk 2 above)
Comment (miyazakh) [reply src/ssl.c:56225]: Agree. Added check ret in the "if" first.

File: src/ssl.c [hunk 7]
```diff
@@ -55992,36 +55994,438 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
  */
 int wolfSSL_CONF_CTX_finish(WOLFSSL_CONF_CTX* cctx)
 {
-    WOLFSSL_STUB("wolfSSL_CONF_CTX_finish");
     (void)cctx;
-    return WOLFSSL_FAILURE;
+    return WOLFSSL_SUCCESS;
+}
+/* 
+ * This comment attempts to describe following definitions ans static functions
```
Comment (cconlon) [inline src/ssl.c:56008]: Let's remove "This comment attempts...", and instead just say "The following definitions and static functions are used for wolfSSL_CONF_cmd() ...".  Also, "ans" should be "and", thanks.

File: src/ssl.c [hunk 8]
```diff
@@ -55992,36 +55994,438 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+ * that are used for wolfSSL_CONF_cmd() to handle command.
+ * 
+ * The following deinitions use for a part of conf_cmds_tbl[] contents.
```
Comment (cconlon) [inline src/ssl.c:56011]: Spelling - "deinitions" should be "definitions"

File: src/ssl.c [hunk 9]
```diff
@@ -55992,36 +55994,438 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+ *  WOLFSSL_CONF_FILE_CMDx  represents command name in configuration file 
+ *  WOLFSSL_CONF_CMDL_CMDx  represents command name on command line
+ * 
+ * The static functions after the definition section process 
+ *                 those FILE or CMDL which are defined in the conf_cmds_tbl.
+ * 
+ * To add a new command handling:
+ *  1. Add new #define to a section of WOLFSSL_CONF_FILE_CMD* and
+ *                                      WOLFSSL_CONF_CMDL_CMD*
+ *  2. Add new statci function after #define section, before
```
Comment (cconlon) [inline src/ssl.c:56021]: Spelling - "statci" should be "static"

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 7 above)
Comment (miyazakh) [reply src/ssl.c]: Changed

File: src/ssl.c (hunk 8 above)
Comment (miyazakh) [reply src/ssl.c]: Fixed

File: src/ssl.c (hunk 9 above)
Comment (miyazakh) [reply src/ssl.c]: Fixed

File: src/ssl.c [hunk 10]
```diff
@@ -56110,36 +56112,436 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
+     /* sanity check */
+    if (cctx == NULL)
+        return WOLFSSL_FAILURE;
+    
+    if (value == NULL) {
+        WOLFSSL_MSG("bad arguments");
+        return ret;
+    }
+    
+    if (!(cctx->flags & WOLFSSL_CONF_FLAG_CERTIFICATE)) {
```
Comment (cconlon) [inline src/ssl.c:56367]: Is it correct for WOLFSSL_CONF_FLAG_CERTIFICATE to be checked here for DH params?

File: src/ssl.c [hunk 11]
```diff
@@ -56110,36 +56112,436 @@ unsigned int wolfSSL_CONF_CTX_set_flags(WOLFSSL_CONF_CTX* cctx,
...
```
Comment (cconlon) [inline src/ssl.c:56319]: Is it correct for WOLFSSL_CONF_FLAG_CERTIFICATE to be checked here for key import?

**Review by cconlon — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 10 above)
Comment (miyazakh) [reply src/ssl.c:56367]: No. The check doesn't need.

File: src/ssl.c (hunk 11 above)
Comment (miyazakh) [reply src/ssl.c:56319]: Yes. It needs to check the flag before importing key.

**Review by cconlon — APPROVED** [review]

**Pull request merged**